Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,719 @@
+    0
+    >>> print(sampler.next())
+    4
+    >>> print(sampler.get_samples(5))
+    [ 0  4  8 12 16]
+
+    Example
+    -------
+    >>> sampler = Sampler.sequential(10)
+    >>> print(sampler.get_samples(5))
+    >>> print(next(sampler))
+    0
+    [0 1 2 3 4]
+    >>> print(sampler.next())
+    1
+
+    Example
+    -------
+    >>> sampler = Sampler.herman_meyer(12)
+    >>> print(sampler.get_samples(16))
+    [ 0  6  3  9  1  7  4 10  2  8  5 11  0  6  3  9]
+
+
+
+    Example
+    --------
+    This example creates a sampler that outputs sequential indices, starting from 1.  
+
+    >>> num_indices=10
+    >>>
+    >>> def my_sampling_function(iteration_number):
+    >>>     return (iteration_number+1)%10
+    >>>    
+    >>> sampler = Sampler.from_function(num_indices=num_indices, function=my_sampling_function)
+    >>> print(list(sampler.get_samples(25)))
+    [1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5]
+
+
+
+    Note
+    -----
+    The optimal choice of sampler depends on the data and the number of calls to the sampler.  Note that a low number of calls to a random sampler won't give an even distribution
+    For a small number of samples (e.g. `<5*num_indices`) the user may wish to consider another sampling method e.g. random without replacement, which, when calling `num_indices` samples is guaranteed to draw each index exactly once.  
+        """
+
+    def __init__(self, num_indices, function,  sampling_type=None, prob_weights=None):
+
+        self._type = sampling_type
+
+        if num_indices == int(num_indices):
+            self._num_indices = num_indices
+        else:
+            raise ValueError('`num_indices` should be an integer. ')
+
+        if callable(function):
+            self._function = function
+        else:
+            raise ValueError('`function` should be an callable function. ')
+
+        if prob_weights is None:
+            prob_weights = [1/num_indices]*num_indices
+        else:
+            if abs(sum(prob_weights)-1) > 1e-6:
+                raise ValueError('The provided prob_weights must sum to one')
+
+            if any(np.array(prob_weights) < 0):
+                raise ValueError(
+                    'The provided prob_weights must be greater than or equal to zero')
+
+        self._prob_weights = prob_weights
+        self._iteration_number = 0
+
+    @property
+    def prob_weights(self):
+        return self._prob_weights
+
+    @property
+    def num_indices(self):
+        return self._num_indices
+
+    @property
+    def current_iter_number(self):
+        return self._iteration_number
+
+    def next(self):
+        """ 
+        Returns a sample from the list of indices `{0, 1, …, N-1}, where N is the number of indices and increments the sampler.
+        """
+
+        out = self._function(self._iteration_number)
+
+        self._iteration_number += 1
+        return out
+
+    def __next__(self):
+        return self.next()
```

Comment:
Do we want to support the use of `sampler.next()`?

Wouldn't it be better to just support `next(sampler)`?